device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/ClientManagerBase.java

```diff
@@ -0,0 +1,220 @@
+package samples.com.microsoft.azure.sdk.iot;
+
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeCallback;
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason;
+import com.microsoft.azure.sdk.iot.device.exceptions.DeviceOperationTimeoutException;
+import com.microsoft.azure.sdk.iot.device.exceptions.TransportException;
+import com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+
+import static com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason.RETRY_EXPIRED;
+import static com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus.DISCONNECTED;
+
+/**
+ * This is the base class for DeviceClientManager and MultiplexingClientManager
+ * The shared logic for reconnection and handling status change callback exists here.
+ */
+@Slf4j
+public abstract class ClientManagerBase implements IotHubConnectionStatusChangeCallback, ConnectionStatusTracker
+{
+    protected static final int SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS = 2;
+
+    // Initialize the connection status as DISCONNECTED
+    protected ConnectionStatus lastKnownConnectionStatus = ConnectionStatus.DISCONNECTED;
+
+    // Tracks connection status of the dependency connection (multiplexing client connection status for example in this case for DeviceClientManager)
+    protected ConnectionStatusTracker dependencyConnectionStatusTracker;
+
+    public abstract void openClient() throws IOException;
+    public abstract void closeClient() throws IOException;
+    public abstract String getClientId();
+
+    // Since the client manager is in charge of handling the connection status callback, this method is a no-op
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext)
+    {
+        return;
+    }
+
+    public void closeNow()
```

Comment:
Why `closeNow()` instead of just `close()`?

Reply:
That is an excellent question.
DeviceClient has two different method signatures, close() and closeNow(). close() has been deprecated 

client managers for both DeviceClient and MultiplexingClient are setup such that any call that you make to them they will delegate to the inner client except for open, close and registerConnectionStatusCallback

Now there are 2 different method signatures on the DeviceClient for closing the connection, and we want to make sure we delegate those calls to the client manager and it knows how to override both of those methods.